src/writer.jl

```diff
@@ -0,0 +1,309 @@
+"""
+write(path::AbstractString, h::Handle) = write(path, h.shapes)
+function write(path::AbstractString, obj; force=false)
+    paths = _shape_paths(path)
+
+    # File existance check
+    if isfile(paths.shp)
+        if force
+            rm(paths.shp)
+        else
+            throw(ArgumentError("File already exists at `$(paths.shp)`. Use `force=true` to write anyway."))
+        end
+    end
+
+    # Handle tabular data
+    if Tables.istable(obj)
+        geomcol = GI.geometrycolumn(obj)
+        geoms = Tables.getcolumn(obj, geomcol)
+        @warn "DBFTables.jl does not yet `write`, so no .dbx file can be written"
+        # DBFTables.Table(obj) # TODO remove geom column
+        # DBFTables.write # TODO DBF write function
+    else
+        geoms = obj
+    end
+
+    # Initialisation
+    shx_indices = IndexRecord[]
+    bytes = 0
+    content_size = 0
+    first_geom = true
+    local mbr, zrange, mrange
+
+    # Open the .shp file as an IO stream
+    io = open(paths.shp, "a+")
+
+    # Write an emtpy header 
+    # We go back later and write it properly later, so we don't have to precalculate everything.
+    dummy_header = Header(;
+        filesize=0, 
+        shapecode=SHAPECODE[Missing],
+        mbr=Rect(0.0, 0.0, 0.0, 0.0),
+        zrange=Interval(0.0, 0.0),
+        mrange=Interval(0.0, 0.0),
+    )
+    bytes += Base.write(io, dummy_header)
+    header_bytes = bytes
+
+    # Detect the shape type code from the first available geometry
+    # There can only be one shape type in the file.
+    # If all values are missing, the shape type is Missing
+    if iterate(skipmissing(geoms)) == nothing
+        trait = Missing
+        hasz = hasm = false
+        shapecode = 0 
+        mbr = Rect(0.0, 0.0, 0.0, 0.0)
+        zrange = Interval(0.0, 0.0)
+        mrange = Interval(0.0, 0.0)
+    else
+        geom1 = first(skipmissing(geoms))
+        GI.isgeometry(geom1) || error("$(typeof(geom1)) is not a geometry")
+        trait = GI.geomtrait(geom1)
+        if trait isa GI.GeometryCollectionTrait 
+            @warn "Geometry Collections or MultiPatch cannot be written using Shapefile.jl"
+            return nothing
+        end
+        hasz = GI.is3d(geom1)
+        hasm = GI.ismeasured(geom1)
+        shapecode = if hasz
+            SHAPECODE[TRAITSHAPE_Z[typeof(trait)]]
+        elseif hasm
+            SHAPECODE[TRAITSHAPE_M[typeof(trait)]]
+        else
+            SHAPECODE[TRAITSHAPE[typeof(trait)]]
+        end
+    end
+
+    # Write the geometry data into io
+    for (num, geom) in enumerate(geoms)
+        (trait === Missing || trait === GI.geomtrait(geom)) || throw(ArgumentError("Shapefiles can only contain geometries of the same type"))
+
+        # One-based record number; Increment from 0, and increment after record.
+        rec_bytes = 0
+        rec_start = bytes
+        calc_rec_bytes = sizeof(Int32) # shape code
+
+        # Write record number
+        bytes += Base.write(io, bswap(Int32(num)))
```

Comment:
This is actually the `hton` from before right?

Reply:
yes I'm actually not sure why they used both `bswap` and `hton` (these are from the original PR) will fix